pty/pty_windows.go

```diff
@@ -134,39 +137,69 @@ func (p *ptyWindows) Resize(height uint16, width uint16) error {
 		Y: int16(height),
 		X: int16(width),
 	})))))
-	if ret != 0 {
+	if windows.Handle(ret) != windows.S_OK {
 		return err
 	}
 	return nil
 }
 
-func (p *ptyWindows) Close() error {
-	p.closeMutex.Lock()
-	defer p.closeMutex.Unlock()
-	if p.closed {
-		return nil
-	}
-	p.closed = true
-
+// closeConsoleNoLock closes the console handle, and sets it to
+// windows.InvalidHandle. It must be called with p.closeMutex held.
+func (p *ptyWindows) closeConsoleNoLock() error {
 	// if we are running a command in the PTY, the corresponding *windowsProcess
 	// may have already closed the PseudoConsole when the command exited, so that
 	// output reads can get to EOF.  In that case, we don't need to close it
 	// again here.
 	if p.console != windows.InvalidHandle {
+		//
+		// Because ClosePseudoConsole has no return value, we could ignore the
+		// error here, but we limit this either S_OK or S_FALSE as a safety-net
+		// in case other values could be returned.
+		// https://docs.microsoft.com/en-us/windows/console/closepseudoconsole
 		ret, _, err := procClosePseudoConsole.Call(uintptr(p.console))
-		if ret < 0 {
-			return xerrors.Errorf("close pseudo console: %w", err)
+		if ret := windows.Handle(ret); ret != windows.S_OK && ret != windows.S_FALSE {
```

Comment:
Review: I contemplated ignoring ret/err here entirely, however, in my testing it seems this function usually returns `S_FALSE` (1). Since I can't find documentation for this behavior, we allow `S_OK` too. Given positive error values, limiting this to negative doesn't make sense to me: https://learn.microsoft.com/en-us/windows/win32/seccrypto/common-hresult-values

Reply:
S_FALSE is a "success" code, so I think it's still fine to check ret<0 for in general.

In this specific case, Windows docs say that there is no return value, so my guess is that the Windows syscall libraries, or perhaps the Windows kernel itself is writing S_FALSE for this procedure to indicate that it's not an explicit "OK" because the call has no return value.

We could just drop the whole error handling block here and cite the docs that say this procedure doesn't return any values.

Reply:
> Values like 0x80004001 are negative when interpreted as a signed 32 bit integer because they are encoded in [two's compliment](https://en.wikipedia.org/wiki/Two%27s_complement)

True, however the return value from `Call` is an uintptr and this `ret` wasn't being cast to 32-bit. As such the `ret < 0` was nonsensical. I don't know enough about Windows syscall internals to know if we can always cast it, that would assume a syscall always returns HRESULT (maybe it does).

> In this specific case, Windows docs say that there is no return value, so my guess is that the Windows syscall libraries, or perhaps the Windows kernel itself is writing S_FALSE for this procedure to indicate that it's not an explicit "OK" because the call has no return value.

Sounds plausible, this value is at least not coming from Go and Windows [does define success as a positive 32-bit integer](https://learn.microsoft.com/en-us/windows/win32/api/winerror/nf-winerror-succeeded).

> We could just drop the whole error handling block here and cite the docs that say this procedure doesn't return any values.

I would still like to keep it, just in case. Unless we can confirm that this syscall can't fail in some way or form.

Reply:
Changed this to `int32(ret) < 0` now, to be more generic and self-documenting via a new `func winerrorFailed(r1 uintptr) bool` function.